### ISSUE TITLE
Should Transformations know coordinate types?

### ISSUE THREAD

**Issue by andyferris** [FugroRoames/CoordinateTransformations.jl#1]
I can see two viable versions of `AbstractTransformation`. The first lets data be pretty free-form:

```julia
abstract AbstractTransformation

immutable MyTransformation <: AbstractTransformation
end

transform(::MyTransformation, x) = (x[1]+1, x[2]+1, x[3]+1)

x = (1.0, 2.0, 3.0)
trans = MyTransformation()
transform(trans, x)
```

Or the types of `x` and the output could be specified by the `Transformation` parameters:

```julia
abstract AbstractTransformation{OutType, InType}

# A fully-specified transformation...
immutable MyTransformation <: AbstractTransformation{NTuple{3,Float64}, NTuple{3,Float64}}
end
# Another transformation, being partly free-form
immutable AnotherTransformation{T} <: AbstractTransformation{T,T}
   rotation::Float64
end

transform(::MyTransformation, x::NTuple{3,Float64}) = (x[1]+1., x[2]+1., x[3]+1.)

x = (1.0, 2.0, 3.0)
trans = MyTransformation()
transform(trans, x)
```

The transform function could somehow check the input type (`typeof(x)`) matches `InType`. If it matches `Vector{InType}` we could have a default vectorized transformation. 

```julia
function transform{OutT,InT}(trans::AbstractTransform{OutT,InT}, v::Vector{InT})
    [transform(trans, x) for x in v]
end
```

If you would like to define a transformation *method* that works for generic input types, you do something like `AnothterTransformation{T}` above. I'm not quite sure where `OutType` is useful, except in the defining the inverse of a transformation, or compositions (and perhaps it will be useful in the context of differentiation also).

**Comment by andyferris** [comment]
I'm leaning towards removing the parameters and letting users do duck-typing, etc.

**Comment by c42f** [comment]
I think we decided the answer is no it shouldn't... so I guess we can close this for now.

**Comment by andyferris** [comment]
Metanote: I'd probably favour letting PRs close issues, when that is how they are to be resolved. Github's automatic issue closing functionality (when you merge commits with e.g. `closes #1` in the [message](https://help.github.com/articles/closing-issues-via-commit-messages/)) is pretty cool after all! :)
